### PR TITLE
71 projectformular add datafields and page 2

### PR DESCRIPTION
**Pull request by Artromi**

### CODE REVIEW

**Review by fabian-urner — CHANGES_REQUESTED** [review]
1. While using the formular it is a little bit confusing to be able to submit the form while beeing on the first page. Maybe it's better to disable the submit button until u reached the last page and validate the input-fields by clicking the next/back buttons

2. The zip-code input-field should be type="text" so the browser can give u hints by typing and there are a lot of countrys with dashes or slashes in the zipcode

3.  after submitting the form, the dialog stays "open". So clicked a few times and submitted a lot of the same projects 😄

**Review by fabian-urner — CHANGES_REQUESTED** [review]
1. Something is not working with the Email and PhoneNr-Validation. It accepts wrong input anyway...

2. If there is a Error with the backend wen run into a infinite loop (also after closing the dialog):
Bildschirmfoto 2024-06-06 um 06 52 49

-> It would be better if the form is shown again with emty fields or better: with the prefilled fields wich are valid and red shaking fields where the error occurred

3. Because of the country-input-field is of type="text" there will be a lot of Backend-Errors if there is a typo... I quickly wrote a route to get all countrys (/api/v1/countrys) -> maybe it's better to have a listbox with all countrys and "Deutschland" is preselected
(⚠️ you have to choose localName in the json-objects wich cames from the backend. If localName is null -> ignore the country)

**Comment by Artromi** [comment]
> 1. Something is not working with the Email and PhoneNr-Validation. It accepts wrong input anyway...
> 2. If there is a Error with the backend wen run into a infinite loop (also after closing the dialog):
>    Bildschirmfoto 2024-06-06 um 06 53 08
> 
> -> It would be better if the form is shown again with emty fields or better: with the prefilled fields wich are valid and red shaking fields where the error occurred
> 
> 3. Because of the country-input-field is of type="text" there will be a lot of Backend-Errors if there is a typo... I quickly wrote a route to get all countrys (/api/v1/countrys) -> maybe it's better to have a listbox with all countrys and "Deutschland" is preselected
>    (⚠️ you have to choose localName in the json-objects wich cames from the backend. If localName is null -> ignore the country)

-> everything should(!) be woring now :D 
-> also: i made an message on page 2 of the formular in case of validation errors on page 2 because i thought its confusing 
     without any indication why nothing got submitted when on page 2 everything is fine
-> easterEgg:  find the  🤠

File: src/stores/CountryStore.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
+//  ### 🤠 ###
```
Comment (fabian-urner) [inline src/stores/CountryStore.js:1]: found it! 🐣

Comment (Artromi) [reply src/stores/CountryStore.js:1]: 💐

File: src/stores/CountryStore.js [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import { defineStore } from 'pinia'
+import { vdbFetchData } from '@/utils/api'
+
+export const useCountryStore = defineStore('CountryStore', {
+  state: () => {
+    return {
+      fetching: false,
+      token: JSON.parse(localStorage.getItem('user-store')).token,
+      countrysPage: null,
```
Comment (fabian-urner) [inline src/stores/CountryStore.js:10]: there is no Pagination build in for countries. So you can just call it countries. (suddenly I realize that it should be “countries” and not “countrys” 🙈) I change the route on the backend as well...

File: src/stores/CountryStore.js [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      countrysArray: []
+    }
+  },
+  actions: {
+    async getCountrys() {
+      // If there's no token, something went wrong
+      if (!this.token) throw Error('VoloDB-ERROR\n🙅‍♀️ ups! not logged in.')
+      try {
+        this.countrysPage = await vdbFetchData(`/countrys`, 'GET', this.token)
+        this.createCountrysArray()
```
Comment (fabian-urner) [inline src/stores/CountryStore.js:20]: you don't need createCountrysArray... becaus you already got it from the backend (-> no pagination)

File: src/stores/CountryStore.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      } catch (error) {
+        console.error('Error fetching countries:', error)
+      } finally {
+        this.fetching = false
```
Comment (fabian-urner) [inline src/stores/CountryStore.js:24]: fetching never set to "true"

File: src/components/ProjectFormular.vue [hunk 5]
```diff
@@ -137,52 +215,106 @@ export default {
...
         } catch (error) {
           console.error(error)
+          
           // Showing error message just for 5 seconds
           this.errorMessage = true
           setTimeout(() => {
             this.errorMessage = false
           }, 5000)
+          // stop fetching to avoid loop:
+          this.projectStore.fetching = false
```
Comment (fabian-urner) [inline src/components/ProjectFormular.vue:300]: this could not be right. If there is a timeout problem we have to handle it directly in the store or check the backend for errors.

File: src/components/VolunteerFormular.vue [hunk 6]
```diff
@@ -226,14 +226,15 @@ export default {
           setTimeout(() => {
             this.errorMessage = false
           }, 5000)
+        // stop fetching to avoid loop:
+         this.volunteerStoreStore.fetching = false
```
Comment (fabian-urner) [inline src/components/VolunteerFormular.vue:230]: this could not be right. If there is a timeout problem we have to handle it directly in the store or check the backend for errors.

**Review by fabian-urner — CHANGES_REQUESTED** [review]
Everything is working like expected! 🎉

## Just two tiny 🪲's:
1. There is no need in the CountryStore 🤠 to make an Array out of Countrys because you already have on from the backend. Further the array-method never clear the array so on reload you have every country in the ListBox twice and so on...
 
Bildschirmfoto 2024-06-10 um 07 55 26

2. wrong overflow on page 2:
Bildschirmfoto 2024-06-10 um 07 51 52



> ### In general:
> As you say: It's confusing to get errors on the second page for errors on the 1st page. When I go back to the 1st page to fix the error, it's confusing to not be able to submit the form just by the fact that I'm on page 1....
The more we deal with this formular-(validation), the more I find it necessary to use a validation-lib for it...
So in conclusion: I'm not very happy with the form as it is. But I think we have to go further 😄

**Review by fabian-urner — APPROVED** [review]
